### PR TITLE
Permanent errors during offset validation should be retried

### LINKED ISSUE #4427: Fluctuations and resets in committed offsets due to temporal network outage
Description
===========
We were testing a network outage scenario where one of three data centers became unavailable and noticed strange fluctuations and resets in committed offsets **after** the data center went back online. I've observed some anomalies in the logs that might be related to it.

## Temporary errors in host resolution that result in the resetting of offsets

>[thrd:main]: test-topic [2]: offset reset (at offset INVALID (leader epoch 6059), broker 1014) to offset BEGINNING (leader epoch -1): Unable to validate offset and epoch: Local: Host resolution failure: Local: Partition log truncation detected

>[thrd:main]: test-topic [19]: offset reset (at offset BEGINNING (leader epoch -1), broker 1014) to offset BEGINNING (leader epoch -1): failed to query logical offset: Local: Host resolution failure

## Suspicious updates of committed offsets

`2023-09-08T00:13:46.246Z` // Started at correct offset
>Consumer in the group "testbench-1000": "[thrd:main]: Partition test-topic [0] start fetching at offset 57842809 (leader epoch 7007)" Code: "FETCH"; SysLevel: Debug; 

`2023-09-08T00:13:50.051Z` // Race condition? Committed offset and leader epoch for partition 0 is from partition 7 (see log below)
>Consumer in the group "testbench-1000": "[thrd:main]: Topic test-topic [0]: stored offset INVALID (leader epoch -1), committed offset 55171745 (leader epoch 6476): not including in commit" Code: "OFFSET"; SysLevel: Debug; 

`2023-09-08T00:13:50.058Z` 

>Consumer in the group "testbench-1000": "[thrd:main]: Topic test-topic [7]: stored offset 55181424 (leader epoch 6476), committed offset 55171745 (leader epoch 6476): setting stored offset 55181424 (leader epoch 6476)  for commit" Code: "OFFSET"; SysLevel: Debug; 

`2023-09-08T00:13:55.056Z` // Back to normal
>Consumer in the group "testbench-1000": "[thrd:main]: Topic test-topic [0]: stored offset 57842809 (leader epoch 7007), committed offset 57842809 (leader epoch 7007): not including in commit" Code: "OFFSET";"

Here is the graph displaying the committed offsets by partitions for that consumer group:
offsets

Please note that on our test bench the probability of encountering a race condition increases, because the Kubernetes pods running the consumer are constantly being throttled.

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): v2.2.0
 - [x] Apache Kafka version: v2.7.2
 - [x] librdkafka client configuration: 
 ```
 auto.offset.reset: earliest
 ```
 - [x] Operating system: Debian 11.7
 - [x] Provide logs (with `debug=..` as necessary) from librdkafka
 - [x] Provide broker log excerpts: N/A
 - [x] Critical issue

### PR DESCRIPTION
**Pull request by emasab**
During offset validation a permanent error like host resolution failure
would cause an offset reset.
This isn't what's expected or what the Java implementation [does](https://github.com/apache/kafka/blob/7872a1ff5b2e9a0fbbe3d71180a97e29f1549d4f/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsForLeaderEpochUtils.java#L113).
Solved by retrying even in case of permanent errors.

Closes #4427.

### CODE REVIEW

File: tests/0139-offset_validation_mock.c [hunk 1]
```diff
@@ -165,10 +166,9 @@ static void do_test_ssl_error_retried(void) {
                                  "bootstrap.servers", bootstraps,
                                  "batch.num.messages", "1", NULL);
 
-        /* Make OffsetForLeaderEpoch fail with the _SSL error */
-        rd_kafka_mock_push_request_errors(mcluster,
-                                          RD_KAFKAP_OffsetForLeaderEpoch, 1,
-                                          RD_KAFKA_RESP_ERR__SSL);
+        /* Make OffsetForLeaderEpoch fail with the corresponding error code */
+        rd_kafka_mock_push_request_errors(
+            mcluster, RD_KAFKAP_OffsetForLeaderEpoch, 1, err);
```
Comment (milindl) [inline tests/0139-offset_validation_mock.c:171]: We ideally should not be sending local errors via the mock broker response, it's not what happens, right?

**Review by milindl — APPROVED** [review]
LGTM given the context modulo one question. 
Not much we can do about it though, I guess?

File: tests/0139-offset_validation_mock.c (hunk 1 above)
Comment (emasab) [reply tests/0139-offset_validation_mock.c:171]: Yeah, but the effect to test would be the same, the SSL error was local too and this seems a simple way to reproduce them.
Other option is a unit test but more complex to set up.

Comment (milindl) [reply tests/0139-offset_validation_mock.c:171]: Sure, go ahead and merge this, looks good to me

**Pull request merged**